In [77]:
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
import seaborn as sns

In [78]:
df_1=pd.read_csv('train.csv')
df_2=pd.read_csv('test.csv')
df_1

Name Platform  Year_of_Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
14571  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
14572               LMA Manager 2007     X360           2006.0        Sports   
14573        Haitaka no Psychedelica      PSV           2016.0     Adventure   
14574               Spirits & Spells      GBA           2003.0      Platform   
14575            Winning Post 8 2016      PSV           2016.0    Simulation   

          Publisher  NA_Sales  EU_Sales  JP_Sales  Critic_Score  Critic_Count  \
0          Nintendo     41.36     28.96      3.77          76.0          51.0   
1          Nintendo     29.08      3.58      6.81           NaN           NaN   
2          Nintendo     15.68     12.76      3.79          82.0          73.0   
3          Nintendo     15.61     10.93      3.28          80.0          73.0   
4          Nintendo     11.27      8.89     10.22           NaN           NaN   
...             ...       ...       ...       ...           ...           ...   
14571    Tecmo Koei      0.00      0.00      0.01           NaN           NaN   
14572   Codemasters      0.00      0.01      0.00           NaN           NaN   
14573  Idea Factory      0.00      0.00      0.01           NaN           NaN   
14574       Wanadoo      0.01      0.00      0.00           NaN           NaN   
14575    Tecmo Koei      0.00      0.00      0.01           NaN           NaN   

      User_Score  User_Count Developer Rating  Global_Sales  
0              8       322.0  Nintendo      E         82.53  
1            NaN         NaN       NaN    NaN         40.24  
2            8.3       709.0  Nintendo      E         35.52  
3              8       192.0  Nintendo      E         32.77  
4            NaN         NaN       NaN    NaN         31.37  
...          ...         ...       ...    ...           ...  
14571        NaN         NaN       NaN    NaN          0.01  
14572        NaN         NaN       NaN    NaN          0.01  
14573        NaN         NaN       NaN    NaN          0.01  
14574        NaN         NaN       NaN    NaN          0.01  
14575        NaN         NaN       NaN    NaN          0.01  

[14576 rows x 15 columns]

### Dataset Info

As a data scientist of a leading decision analysis firm you are required to predict the potential global user of the game based on the data provided by the customer so that they can plan their global launch. 

Name - Game name
Platform - Running platform
Year of release
Genre - Game Genre
Publisher
Critic_score - Aggregate score compiled by Metacritic staff
Criticcount - The number of critics used in coming up with the Critic Score
User_score - Score by Metacritic’s subscribers
Usercount - Number of users who gave the user score
Developer - Party responsible for creating the game
Rating - The ESRB ratings
NA_Sales - Sales in North America (in millions of units)
EU_Sales - Sales in the European Union (in millions of units)
JP_Sales - Sales in Japan (in millions of units)
Global_Sales - Total sales in the world (in millions of units)

In [79]:
# Data Information
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14576 entries, 0 to 14575
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             14574 non-null  object 
 1   Platform         14576 non-null  object 
 2   Year_of_Release  14344 non-null  float64
 3   Genre            14574 non-null  object 
 4   Publisher        14527 non-null  object 
 5   NA_Sales         14576 non-null  float64
 6   EU_Sales         14576 non-null  float64
 7   JP_Sales         14576 non-null  float64
 8   Critic_Score     7217 non-null   float64
 9   Critic_Count     7217 non-null   float64
 10  User_Score       8760 non-null   object 
 11  User_Count       6796 non-null   float64
 12  Developer        8829 non-null   object 
 13  Rating           8704 non-null   object 
 14  Global_Sales     14576 non-null  float64
dtypes: float64(8), object(7)
memory usage: 1.7+ MB


#### so here 7 object datatype columns and 8 numerical datatype columns are there and shape is 14576 rows and 15 columns

### STATISTICAL Summary

In [80]:
# numerical features
df_1.describe().T

count         mean         std      min      25%      50%  \
Year_of_Release  14344.0  2006.437117    5.955664  1980.00  2003.00  2007.00   
NA_Sales         14576.0     0.295577    0.866491     0.00     0.00     0.10   
EU_Sales         14576.0     0.163957    0.536354     0.00     0.00     0.03   
JP_Sales         14576.0     0.085659    0.329646     0.00     0.00     0.00   
Critic_Score      7217.0    69.676043   13.773391    13.00    61.00    71.00   
Critic_Count      7217.0    27.312041   19.464196     3.00    12.00    22.00   
User_Count        6796.0   172.513096  576.609716     4.00    11.00    27.00   
Global_Sales     14576.0     0.598972    1.647699     0.01     0.05     0.22   

                     75%       max  
Year_of_Release  2010.00   2020.00  
NA_Sales            0.28     41.36  
EU_Sales            0.13     28.96  
JP_Sales            0.03     10.22  
Critic_Score       80.00     98.00  
Critic_Count       38.00    113.00  
User_Count         89.00  10665.00  
Global_Sales        0.55     82.53

#### so here we get the info about total count,mean,standard deviation ,maximum,minimumand many more about the numerical features of the data

In [81]:
# categorical features
df_1.describe(include=object).T

count unique                          top  freq
Name        14574  10327  Need for Speed: Most Wanted    11
Platform    14576     31                          PS2  1935
Genre       14574     12                       Action  2974
Publisher   14527    552              Electronic Arts  1267
User_Score   8760     96                          tbd  1964
Developer    8829   1577                      Ubisoft   171
Rating       8704      8                            E  3460

In [82]:
# conclusion

In [83]:
#sns.pairplot(df_1)

In [84]:
 # Checking the null values
nulls=df_1.isnull().sum()
null=pd.DataFrame({'count':nulls,'percent':(nulls/df_1.shape[0])*100})
null.sort_values('percent',ascending=False) 
    

count    percent
User_Count        7780  53.375412
Critic_Score      7359  50.487102
Critic_Count      7359  50.487102
Rating            5872  40.285401
User_Score        5816  39.901207
Developer         5747  39.427827
Year_of_Release    232   1.591658
Publisher           49   0.336169
Name                 2   0.013721
Genre                2   0.013721
Platform             0   0.000000
NA_Sales             0   0.000000
EU_Sales             0   0.000000
JP_Sales             0   0.000000
Global_Sales         0   0.000000

In [85]:
df_1.Name.mode()

0    Need for Speed: Most Wanted
dtype: object

In [86]:
df_1.Name.fillna('Need for Speed: Most Wanted',inplace=True)

In [87]:
df_1.Genre.mode()

0    Action
dtype: object

In [88]:
df_1.Genre.fillna('Action',inplace=True)

In [89]:
 # Checking the null values
nulls=df_1.isnull().sum()
null=pd.DataFrame({'count':nulls,'percent':(nulls/df_1.shape[0])*100})
null.sort_values('percent',ascending=False) 
    

count    percent
User_Count        7780  53.375412
Critic_Score      7359  50.487102
Critic_Count      7359  50.487102
Rating            5872  40.285401
User_Score        5816  39.901207
Developer         5747  39.427827
Year_of_Release    232   1.591658
Publisher           49   0.336169
Name                 0   0.000000
Platform             0   0.000000
Genre                0   0.000000
NA_Sales             0   0.000000
EU_Sales             0   0.000000
JP_Sales             0   0.000000
Global_Sales         0   0.000000

In [90]:
df_1.Publisher.mode()

0    Electronic Arts
dtype: object

In [91]:
df_1.Publisher.fillna('Electronic Arts',inplace=True)

In [92]:
df_1.Year_of_Release.median()

2007.0

In [93]:
df_1['Year_of_Release'].fillna(df_1['Year_of_Release'].mode()[0],inplace=True)

In [94]:
 # Checking the null values
nulls=df_1.isnull().sum()
null=pd.DataFrame({'count':nulls,'percent':(nulls/df_1.shape[0])*100})
null.sort_values('percent',ascending=False) 
    

count    percent
User_Count        7780  53.375412
Critic_Score      7359  50.487102
Critic_Count      7359  50.487102
Rating            5872  40.285401
User_Score        5816  39.901207
Developer         5747  39.427827
Name                 0   0.000000
Platform             0   0.000000
Year_of_Release      0   0.000000
Genre                0   0.000000
Publisher            0   0.000000
NA_Sales             0   0.000000
EU_Sales             0   0.000000
JP_Sales             0   0.000000
Global_Sales         0   0.000000

In [95]:
#Games publishers give developers the money upfront to make a game.
#They strike up a deal with a studio (game development company) to make a game according to an agreed brief. The publisher 
#then works closely with the game producer at the studio to make sure the game's being made as agreed, to budget and on time.


In [20]:
#dv_pvt=df_1.pivot_table(values='Global_Sales',index=['Genre'],aggfunc=np.mean)
#dv_pvt=df_1.pivot_table(values='Developer',index=['Publisher'],aggfunc=stats.mode(df_1['Developer']))
#dv_pvt

In [21]:
df_1

Name Platform  Year_of_Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
14571  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
14572               LMA Manager 2007     X360           2006.0        Sports   
14573        Haitaka no Psychedelica      PSV           2016.0     Adventure   
14574               Spirits & Spells      GBA           2003.0      Platform   
14575            Winning Post 8 2016      PSV           2016.0    Simulation   

          Publisher  NA_Sales  EU_Sales  JP_Sales  Critic_Score  Critic_Count  \
0          Nintendo     41.36     28.96      3.77          76.0          51.0   
1          Nintendo     29.08      3.58      6.81           NaN           NaN   
2          Nintendo     15.68     12.76      3.79          82.0          73.0   
3          Nintendo     15.61     10.93      3.28          80.0          73.0   
4          Nintendo     11.27      8.89     10.22           NaN           NaN   
...             ...       ...       ...       ...           ...           ...   
14571    Tecmo Koei      0.00      0.00      0.01           NaN           NaN   
14572   Codemasters      0.00      0.01      0.00           NaN           NaN   
14573  Idea Factory      0.00      0.00      0.01           NaN           NaN   
14574       Wanadoo      0.01      0.00      0.00           NaN           NaN   
14575    Tecmo Koei      0.00      0.00      0.01           NaN           NaN   

      User_Score  User_Count Developer Rating  Global_Sales  
0              8       322.0  Nintendo      E         82.53  
1            NaN         NaN       NaN    NaN         40.24  
2            8.3       709.0  Nintendo      E         35.52  
3              8       192.0  Nintendo      E         32.77  
4            NaN         NaN       NaN    NaN         31.37  
...          ...         ...       ...    ...           ...  
14571        NaN         NaN       NaN    NaN          0.01  
14572        NaN         NaN       NaN    NaN          0.01  
14573        NaN         NaN       NaN    NaN          0.01  
14574        NaN         NaN       NaN    NaN          0.01  
14575        NaN         NaN       NaN    NaN          0.01  

[14576 rows x 15 columns]

In [22]:
#pd.options.display.max_rows=None

In [23]:
d_index=df_1[df_1['Developer'].isnull()].index
d_index

Int64Index([    1,     4,     5,     9,    10,    12,    18,    20,    21,
               22,
            ...
            14562, 14564, 14568, 14569, 14570, 14571, 14572, 14573, 14574,
            14575],
           dtype='int64', length=5747)

In [24]:
#df_1['Publisher'].fillna(df_1['Publisher'].mode()[0],inplace=True)

In [25]:
#df_1['Developer'].fillna(df_1['Developer'].mode()[0],inplace=True)

In [26]:
df_num=df_1.select_dtypes(exclude='object')
df_num.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14576 entries, 0 to 14575
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year_of_Release  14576 non-null  float64
 1   NA_Sales         14576 non-null  float64
 2   EU_Sales         14576 non-null  float64
 3   JP_Sales         14576 non-null  float64
 4   Critic_Score     7217 non-null   float64
 5   Critic_Count     7217 non-null   float64
 6   User_Count       6796 non-null   float64
 7   Global_Sales     14576 non-null  float64
dtypes: float64(8)
memory usage: 911.1 KB


In [28]:
df_1.User_Score.unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '0.6', '1.4', '0.9', '1', '1.3', '0.2', '9.7'],
      dtype=object)

In [99]:
df_1['User_Score'].replace('tbd',None,inplace=True)

In [100]:
df_1['User_Score'].fillna(df_1['User_Score'].median(),inplace=True)

In [101]:
 # Checking the null values
nulls=df_1.isnull().sum()
null=pd.DataFrame({'count':nulls,'percent':(nulls/df_1.shape[0])*100})
null.sort_values('percent',ascending=False) 
    

count    percent
User_Count        7780  53.375412
Critic_Score      7359  50.487102
Critic_Count      7359  50.487102
Rating            5872  40.285401
Developer         5747  39.427827
Name                 0   0.000000
Platform             0   0.000000
Year_of_Release      0   0.000000
Genre                0   0.000000
Publisher            0   0.000000
NA_Sales             0   0.000000
EU_Sales             0   0.000000
JP_Sales             0   0.000000
User_Score           0   0.000000
Global_Sales         0   0.000000

In [31]:
#from scipy.stats import mode

#Determing the mode for each
#Developer_mode = df_1.pivot_table(values='Developer', columns=['Genre'],aggfunc=(lambda x:mode(x).mode[0]) ).T
##print('Mode for each Oue:')
#Developer_mode

In [32]:
#for i in df_1.Rating:
#    print(i)

In [33]:
#df_r=df_1['Rating'].str.split('',2,expand=True).apply(lambda x:x[0])
#df_r

In [108]:
from scipy.stats import mode
r_bool=df_1['Rating'].isnull().sum()
ratings_mode = df_1.pivot_table(values='Rating', index=[]'Genre',aggfunc=(lambda x:mode(x).mode[0]) )
ratings_mode  
   

Rating
Genre              
Action            T
Adventure         E
Fighting          T
Misc              E
Platform          E
Puzzle            E
Racing            E
Role-Playing      T
Shooter           M
Simulation        E
Sports            E
Strategy          T

In [103]:
r_bool=df_1['Rating'].isnull()
r_bool

0        False
1         True
2        False
3        False
4         True
         ...  
14571     True
14572     True
14573     True
14574     True
14575     True
Name: Rating, Length: 14576, dtype: bool

In [109]:
df_1.loc[r_bool,'Rating'] = df_1.loc[r_bool,'Genre'].apply(lambda x: ratings_mode.loc[x])

AttributeError: 'str' object has no attribute 'apply'

In [37]:
#df_1['Rating'] = df_1['Rating'].fillna(df_1['Rating'].mode()[0])

In [38]:
df_1.Developer.unique()

array(['Nintendo', 'Ubisoft', 'Good Science Studio', ...,
       'Big Red Software', 'Atomic Games', 'Interchannel-Holon'],
      dtype=object)

In [39]:
df_1.Rating.isnull().sum()

0

In [40]:
#df_1.loc[d_index,'Developer']=df_1.loc[d_index,'Publisher'].apply(lambda x:dv_pvt[x])
#bm_data.loc[iw_bool,'Item_Weight'] = bm_data.loc[iw_bool,'Item_Identifier'].apply(lambda x: item_avg_weight.loc[x])

In [41]:
df_1['Critic_Count'].value_counts()

4.0      259
5.0      222
9.0      214
11.0     213
7.0      207
        ... 
103.0      1
102.0      1
106.0      1
3.0        1
104.0      1
Name: Critic_Count, Length: 106, dtype: int64

In [42]:
df_1['Critic_Count'].unique()

array([ 51.,  nan,  73.,  65.,  41.,  80.,  64.,  63.,  45.,  33.,  50.,
        58.,  62.,  37.,  54.,  81.,  89.,  21., 100.,  39.,  56.,  66.,
        57.,  86.,  74.,  16.,  82.,  67.,  29.,  15.,  70.,  99.,  20.,
        87.,  23.,  24.,  10.,  27.,  42.,  91.,  53.,  75.,  46.,  84.,
        38.,  12.,  90.,  97.,  44., 105.,  77.,  68.,  28.,  94.,  96.,
        61.,  22.,  88.,  47.,  98.,  85.,  13.,   8.,  95.,  17.,  35.,
        19.,  48.,  14., 113.,  83.,  72.,  30.,  69.,  76.,  18.,  34.,
        40., 101.,   9.,  32.,  78.,  36.,  25.,  59.,  31.,  79.,  26.,
        49.,  60.,  11.,  43., 107.,   7.,  71.,  55.,  52.,  92.,   4.,
        93.,   5.,   6., 103., 102., 106.,   3., 104.])

In [43]:
df_1.Rating.isnull().sum()

0

In [74]:
from scipy.stats import mode
critic_count=df_1.pivot_table(values='Critic_Count',index=['Genre'],aggfunc=(lambda x:mode(x).mode[0]))
critic_count

Critic_Count
Genre                     
Action                 4.0
Adventure              4.0
Fighting               4.0
Misc                   4.0
Platform               4.0
Puzzle                 4.0
Racing                 4.0
Role-Playing           4.0
Shooter                4.0
Simulation             4.0
Sports                 4.0
Strategy               4.0

In [75]:
c_bool=df_1['Critic_Count'].isnull()
c_bool

0        False
1        False
2        False
3        False
4        False
         ...  
14571    False
14572    False
14573    False
14574    False
14575    False
Name: Critic_Count, Length: 14576, dtype: bool

In [106]:
df_1.loc[c_bool,'Critic_Count']=df_1.loc[c_bool,'Genre'].apply(lambda x:critic_count[x])

In [71]:
df_1['Critic_Count']=df_1['Critic_Count'].fillna(df_1['Critic_Count'].mode()[0])

In [72]:
df_1['Critic_Score']=df_1['Critic_Score'].fillna(df_1['Critic_Score'].median())

In [ ]:
df_1['User_count']